Pursuing the pheromone trails

Jonah-gr opened the pull request:
The ants now follow the pheromone trails and this works as follows:
1. the position of the respective ant in the grid is calculated
2. the pheromone field with the highest value in a radius `Ant.search_radius` around the ant is calculated
3. the position of the previously calculated field is scaled again in the coordinate system
4. the direction vector from the ant to the vector of the pheromone is calculated
5. this vector is passed in `Simulation.next_epoch()`  to `Ant.move(pheromone_direction)` and is added to `self.direction` (if available).

`Ant.search_radius` is adjustable via the gui and its default value is 3.
I added the 2 methods I needed for this in simulation.py. This seemed the easiest.

The whole thing works, but how well ant trails are actually formed depends heavily on the parameters set (size of the pheromone grid, size of the world, `reducing_factor` and `zero_threshold` of `Pheromone.reduce_pheromone()`, step size of the ants etc.).
I think there is still room for improvement in the parameters, but I think this is a good basis. Let me know what you think:)

(I hope nobody minds that I worked on it now)

[![Image from Gyazo](https://i.gyazo.com/c60f0170e75e60c0080c7616ef966890.gif)](https://gyazo.com/c60f0170e75e60c0080c7616ef966890)

tis22 commented on resources/simulation.py:
```diff
@@ -46,12 +46,14 @@ def next_epoch(self):
                 
                 idx_row, idx_col = self.map_ant_coordinates_to_pheromone_index(ant_coordinates = ant.coordinates,
                                                                                colony = colony)
+                colony.pheromone.reduce_pheromones(0.9999, 0.0001)
```
This should take place after all ants (of the colony) have left their pheromone.
```suggestion
                colony.pheromone.leave_pheromone(pos = (idx_row, idx_col), pheromone_status = ant.pheromone_status)
           colony.pheromone.reduce_pheromones(0.9999, 0.0001)
```

tis22 commented:
Looks good, I've almost reached the same with my implementation.
You took my ideas and suggestions into account, implemented them well and continued to make improvements according to your ideas! 😄 

Next time it would be better to communicate beforehand because you've already contributed a lot and we should try to have everyone contribute an equal part to the code. 😅 

Furthermore I've encountered a problem in your implementation.
The ants searching for the food and finding the food.
After that they don't reach the colony and circulate around one pheromone-field.

I guess this happens: 
If an ant stands in the middle (field with highest pheromone-level) it walks to another field next to it, in the next epoch it goes back to its position one epoch before (because it has the highest pheromone-level).
This maybe happens because no ant has already reached the colony or it has something to to with the step_size.

![ants](https://github.com/heyitsalina/ant_search_algorithm/assets/118851752/d318ba56-e7ec-4dc9-9533-f3009af0082b)


After several more epochs:
![Screenshot 2024-02-08 204722](https://github.com/heyitsalina/ant_search_algorithm/assets/118851752/4c63d0f9-90fa-41b1-9504-21184d4fec37)


Another problem happens if I press the clear-button.
The background disappears then.
![Screenshot 2024-02-08 204918](https://github.com/heyitsalina/ant_search_algorithm/assets/118851752/22c3d318-c936-4892-b7af-2e51ae94a032)

Jonah-gr commented:
The problem that the ants eventually only run in circles is what I meant by the tricky parameter settings. This does not always happen, but unfortunately very often. Off the top of my head, I wouldn't know how to get around this, but somehow it's part of the algorithm, I think. 

I've also noticed that the background sometimes stops loading. It seems to happen especially if you wait a long time before starting the simulation (again). I will try again to fix the error somehow, but I strongly suspect that it is due to Kivy.

I know that Anas and you are also working on it somehow but haven't made much progress in a while.
I hope I was able to give you some inspiration with this :)

AlhaririAnas commented:
I worked on it also and, as seen in the GIF, the ants can now navigate based on pheromones. However, my approach is not **optimal** and has two main drawbacks: 

1. Firstly, 1-3 ants should remain in the colony to "_help within the colony, for example with storing food_" . This would also create continuous pheromone trails within the colony, aiding other ants in orientation.
2. Secondly, the ants currently move to the center of a pheromone field, meaning they can't reach the world's boundaries. A possible solution might be to enlarge the pheromone grid so that the centers of the outermost pheromone fields align exactly with the world's edges.

My approach serves its purpose but is not the most efficient! I will update my current pull request later so you can see exactly what I've done.

![UnbenanntesVideoMitClipchamperstellt2-ezgif com-video-to-gif-converter](https://github.com/heyitsalina/ant_search_algorithm/assets/147917604/e96d7530-885f-4b94-b45c-293d1905d486)

tis22 commented:
To 1.:
I guess this would not be the solution we are looking for neither this leads to a better outcome because if ants are too far away they would not notice these pheromones (but in the real world there would be Asien ants in the colony always anyway though…). Furthermore if we implement this we would need ants for both types of pheromones there.


I guess a good idea could be to have a „timer“ when an ant moves back to the colony (coordinates)/ it‘s direction. This could be done by checking how many epochs have passed and after this amount the ant would go back (maybe this number could be some kind of random for each ant so that it reflects their own „will“ better?).

tis22 commented:
The new commits enable my proposed idea: the ants will return to the colony after a certain number of epochs regardless of the pheromones. After that they orient themselves again to the pheromone-level.

Jonah-gr commented:
> The new commits enable my proposed idea: the ants will return to the colony after a certain number of epochs regardless of the pheromones. After that they orient themselves again to the pheromone-level.

Interesting feature, but should we really implement it? Or should it at least be possible to switch it on/off? Because of course that's not really a realistic representation.

tis22 commented:
I thought of it as a solution of the circulating around one pheromone-field and I wanted to test how we can work around the problem further.
The commits can be reverted or just removed from the code if we do not need them.

honnigmelone commented:
Hey, I think it's realistic behavior if ants can also wind up in those pheromone circles ("Ameisenmühle"). So we should only add that as a switch on/off feature. But it is still an interesting idea!

Jonah-gr commented:
I have now added a parameter `pheromone_influence` which indicates how strongly the direction of movement of the ants depends on the direction of the highest perceived pheromone.

I also fixed the user input for the `search_radius`.


How should we proceed now? I think we should merge something soon.

honnigmelone reviewed (approved):
I agreee and I think it's a good implementation now :) Great work!! I would also be fine with merging this now.

heyitsalina commented:
yeah i agree with nikita, it looks great! :) would not have anything against merging since we need something soon

AlhaririAnas commented:
Ok, we can take this approach and merge it with our main if everyone agrees... nice job :)
